citro3d/src/math/matrix.rs

```diff
@@ -1,102 +1,87 @@
-
-        // This constructor validates, at compile time, that the
-        // constructed matrix is 3xN or 4xN matrix, where 0 < N ≤ 4.
-        // We put this struct in a submodule to enforce that nothing creates
-        // a Matrix without calling this constructor.
-        #[allow(clippy::let_unit_value)]
-        pub(crate) fn new(value: citro3d_sys::C3D_Mtx) -> Self {
-            let () = Self::ROW_SIZE;
-            let () = Self::COLUMN_SIZE;
-            Self(value)
-        }
-
-        pub(crate) fn as_raw(&self) -> *const citro3d_sys::C3D_Mtx {
-            &self.0
-        }
+use super::{CoordinateOrientation, FVec3, FVec4};
 
-        pub(crate) fn into_raw(self) -> citro3d_sys::C3D_Mtx {
-            self.0
-        }
+/// A 4x4 row-major matrix of `f32`s.
+///
+/// # Layout details
+/// Rows are actually stored as WZYX in memory. There are helper functions
+/// for accessing the rows in XYZW form. The `Debug` implementation prints
+/// the shows in WZYX form
+///
+/// It is also guaranteed to have the same layout as [`citro3d_sys::C3D_Mtx`]
+#[doc(alias = "C3D_Mtx")]
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct Matrix4(citro3d_sys::C3D_Mtx);
 
-        pub(crate) fn as_mut(&mut self) -> *mut citro3d_sys::C3D_Mtx {
-            &mut self.0
-        }
+impl Matrix4 {
+    /// Construct a Matrix4 from the cells
+    ///
+    /// # Note
+    /// This expects rows to be in WZYX order
+    pub fn from_cells_wzyx(cells: [f32; 16]) -> Self {
+        Self(citro3d_sys::C3D_Mtx { m: cells })
+    }
+    /// Construct a Matrix4 from its rows
+    pub fn from_rows(rows: [FVec4; 4]) -> Self {
+        Self(citro3d_sys::C3D_Mtx {
+            // Safety: FVec is repr(transparent)
+            r: unsafe { core::mem::transmute::<_, [citro3d_sys::C3D_FVec; 4]>(rows) },
+        })
+    }
+    /// Create a new matrix from a raw citro3d_sys one
+    pub fn from_raw(value: citro3d_sys::C3D_Mtx) -> Self {
+        Self(value)
+    }
 
-        /// Trim the matrix down to only the rows and columns we care about,
-        /// since the inner representation is always 4x4.
-        ///
-        /// NOTE: this probably shouldn't be used in hot paths since it copies
-        /// the underlying storage. For some use cases slicing might be better,
-        /// although the underlying slice would always contain extra values for
-        /// matrices smaller than 4x4.
-        pub(crate) fn as_rows(&self) -> [[f32; N]; M] {
-            let rows = unsafe { self.0.r }.map(|row| -> [f32; N] {
-                // Rows are stored in WZYX order, so we slice from back to front.
-                // UNWRAP: N ≤ 4, so slicing to a smaller array should always work
-                unsafe { row.c[(4 - N)..].try_into() }.unwrap()
-            });
-
-            // UNWRAP: M ≤ 4, so slicing to a smaller array should always work
-            rows[..M].try_into().unwrap()
-        }
+    pub fn as_raw(&self) -> &citro3d_sys::C3D_Mtx {
```

Comment:
`as_raw` because if we want to pass to the C API we need a reference to the underlying `C3D_Mtx`. If we're just passing constant pointers which never mutate (which I think might be the case actually) this is fine, but its a potential footgun that I didn't want to think too hard about. I can make it only `as_raw(self) -> citro3d_sys::C3D_Mtx` if you like?

Reply:
I guess I don't have a strong opinion either way on this, I would think that `as_raw` might be more "universal" since you could just deref to make a copy?